src/pip/_internal/index.py

```diff
@@ -939,15 +939,25 @@ def _get_encoding_from_headers(headers):
     return None
 
 
-_CLEAN_LINK_RE = re.compile(r'[^a-z0-9$&+,/:;=?@.#%_\\|-]', re.I)
-
-
 def _clean_link(url):
     # type: (str) -> str
     """Makes sure a link is fully encoded.  That is, if a ' ' shows up in
     the link, it will be rewritten to %20 (while not over-quoting
     % or other characters)."""
-    return _CLEAN_LINK_RE.sub(lambda match: '%%%2x' % ord(match.group(0)), url)
+    # Split url into parts.
+    result = urllib_parse.urlparse(url)
+    if result.netloc == "":
```

Comment:
Since it's not obvious what this means / translates to, I would elaborate that this covers / is equivalent to the file URL case (`file://`).

Reply:
Done. I have added more comments.